idx/metric_definition.go

```diff
@@ -0,0 +1,352 @@
+	bld := strings.Builder{}
+
+	bld.WriteString(t.Key)
+	bld.WriteString("=")
+	bld.WriteString(t.Value)
+
+	return bld.String()
+}
+
+// TagKeyValues stores a slice of all of the Tag Key/Value pair combinations for a MetricDefinition
+type TagKeyValues []TagKeyValue
+
+// Strings returns a slice containing all of the Tag Key/Value pair combinations for a MetricDefinition.
+// Each item in the slice is in the form of 'key=value'
+func (t TagKeyValues) Strings() []string {
+	tags := make([]string, len(t))
+	for i, tag := range t {
+		tags[i] = tag.String()
+	}
+	return tags
+}
+
+// Helper functions to sort TagKeyValues
+func (t TagKeyValues) Len() int           { return len(t) }
+func (t TagKeyValues) Swap(i, j int)      { t[i], t[j] = t[j], t[i] }
+func (t TagKeyValues) Less(i, j int) bool { return t[i].Key < t[j].Key }
+
+// MetricDefinition stores information which identifies a single metric
+type MetricDefinition struct {
+	Id    schema.MKey
+	OrgId uint32
+	// using custom marshalling for MetricName
+	// if there is another way we should explore that
+	Name       MetricName `msg:"name,extension"`
+	Interval   int
+	Unit       string
+	mtype      mType
+	Tags       TagKeyValues
+	LastUpdate int64
+	Partition  int32
+}
+
+// NameWithTags returns a string version of the MetricDefinition's name with
+// all of its tagsin the form of 'name;key1=value1;key2=value2;key3=value3'
+func (md *MetricDefinition) NameWithTags() string {
+	bld := strings.Builder{}
+
+	md.Name.string(&bld)
+	sort.Sort(TagKeyValues(md.Tags))
+	for _, tag := range md.Tags {
+		if tag.Key == "name" {
+			continue
+		}
+		bld.WriteString(";")
+		bld.WriteString(tag.String())
+	}
+	return bld.String()
+}
+
+// SetMType translates a string into a uint8 which is used to store
+// the actual metric type. Valid values are 'gauge', 'rate', 'count',
+// 'counter', and 'timestamp'.
+func (md *MetricDefinition) SetMType(mtype string) {
+	switch mtype {
+	case "gauge":
+		md.mtype = MTypeGauge
+	case "rate":
+		md.mtype = MTypeRate
+	case "count":
+		md.mtype = MTypeCount
+	case "counter":
+		md.mtype = MTypeCounter
+	case "timestamp":
+		md.mtype = MTypeTimestamp
+	default:
+		// for values "" and other unknown/corrupted values
+		md.mtype = MTypeUndefined
+	}
+}
+
+// Mtype returns a string version of the current MType
+func (md *MetricDefinition) Mtype() string {
+	switch md.mtype {
+	case MTypeGauge:
+		return "gauge"
+	case MTypeRate:
+		return "rate"
+	case MTypeCount:
+		return "count"
+	case MTypeCounter:
+		return "counter"
+	case MTypeTimestamp:
+		return "timestamp"
+	default:
+		// case of MTypeUndefined and also default for unknown/corrupted values
+		return ""
+	}
+}
+
+// SetUnit takes a string, interns it in an object store
+// and then uses it to store the unit.
+func (md *MetricDefinition) SetUnit(unit string) {
+	sz, err := IdxIntern.AddOrGetString([]byte(unit))
+	if err != nil {
+		log.Errorf("idx: Failed to intern Unit %v. %v", unit, err)
+		md.Unit = unit
```

Comment:
this should return before leaving the if block

Reply:
Yeah, I really need to just implement handling AddOrGet errors the same way across the entire PR. Either we try to recover, log it, and continue, or we log it and fail.